source/vibe/core/net.d

```diff
@@ -586,6 +586,41 @@ mixin(tracer);
 		return m_context.readBuffer.length > 0;
 	}
 
+	enum WaitForDataAsyncStatus {
+		noMoreData,
+		dataAvailable,
+		waiting
+	}
+
+	WaitForDataAsyncStatus waitForDataAsync(void delegate(bool) @safe read_callback, Duration timeout = Duration.max)
+	{
+mixin(tracer);
+		import vibe.core.core : runTask, setTimer, createTimer;
+
+		if (!m_context) {
+			runTask(read_callback, false);
+			return WaitForDataAsyncStatus.noMoreData;
+		}
+		if (m_context.readBuffer.length > 0) {
+			runTask(read_callback, true);
+			return WaitForDataAsyncStatus.dataAvailable;
+		}
+
+		if (timeout <= 0.seconds) {
+			auto rs = waitForData(0.seconds);
+			auto mode = rs ? WaitForDataAsyncStatus.dataAvailable : WaitForDataAsyncStatus.noMoreData;
+			return mode;
+		}
+
+		auto tm = setTimer(timeout, { eventDriver.sockets.cancelRead(m_socket);
+				runTask(read_callback, false); });
+		eventDriver.sockets.read(m_socket, m_context.readBuffer.peekDst(), IOMode.once,
+				(sock, st, nb) { tm.stop(); if(st != IOStatus.ok) runTask(read_callback, false);
```

Comment:
The two lines from above
```
assert(m_context.readBuffer.length == 0);
m_context.readBuffer.putN(nbytes);``
```

should be executed right after the `tm.stop()`, so that the read data is actually committed to the read buffer. The `if` condition would then ideally be `m_context.readBuffer.length > 0` instead of `st == IOStatus.ok`, because partial data could have been read before the error was raised (not a realistic case, since `IOMode.once` is used, but handling it this way should be slightly more robust).